sdk/test-utils/recorder-new/MIGRATION.md

```diff
@@ -0,0 +1,160 @@
+
+| Script              | Before migration                                                                                                                                                                         | After migration                                                                                                                        |
+| :------------------ | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :------------------------------------------------------------------------------------------------------------------------------------- |
+| `unit-test:browser` | `karma start --single-run`                                                                                                                                                               | `dev-tool run test:browser`                                                                                                            |
+| `unit-test:node`    | `mocha -r esm -r ts-node/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 1200000 --full-trace --exclude \"test/**/browser/*.spec.ts\" \"test/**/*.spec.ts\"` | `dev-tool run test:node-ts-input --mocha=\"--timeout 1200000 --full-trace --exclude 'test/**/browser/*.spec.ts' 'test/**/*.spec.ts'\"` |
+
+## Initializing the recorder
+
+The approach taken to initialize the recorder depends on whether the SDK being tested uses Core v1 (`core-http`) or Core v2 (`core-rest-pipeline`). If your SDK is on Core v2, read on. If you're still on Core v1, jump to the section on Core v1 below.
+
+### For Core v2 SDKs
+
+The recorder is implemented as a custom policy which should be attached to your client's pipeline. Firstly, initialize the recorder:
+
+```ts
+let recorder: TestProxyHttpClient;
+
+/*
+ * Note the use of function() instead of the arrow syntax. We need access to `this` so we
+ * can pass test information from Mocha to the recorder.
+ */
+beforeEach(function(this: Context) {
+  recorder = new TestProxyHttpClient(this.currentTest);
+});
+```
+
+To enable the recorder, you should then initialize your SDK client as normal and use the recorder's `configureClient` method. This method will attach the necessary policies to the client for recording to be enabled.
+
+```ts
+const client = /* ... initialize your client as normal ... */;
+// recorderHttpPolicy is provided as an export from the test-recorder-new package.
+recorder.configureClient(client);
+```
+
+### For Core v1 SDKs
+
+The recorder library provides a custom `HttpClient` that is then passed to the SDK. This client needs to be initialized as follows:
+
+```ts
+let recorder: TestProxyHttpClientCoreV1;
```

Comment:
I’m slightly confused by the name of this variable. The previous recorder was a complex entity that went way beyond an `httpClient`. This variable at first glance seems to be analogous to the previous recorder, but it really intends to be passed as an `httpClient`. What if this variable is called `httpClient`?

Reply:
Or perhaps the HTTP client could be a specific property? Like `recorder.httpClient`?

Reply:
(If you like this idea, this could be a separate issue, to avoid blocking this PR)

Reply:
I like your second idea, but definitely out of scope here, and I suspect it would be a significant refactor. @HarshaNalluru may also have thoughts.

Reply:
It would have been really hard to bring this up before because it wasn’t as easy to see the recorder in context. Please consider this last phase (before migrating clients to the new recorder) to be a critical step to factor in design feedback. After the packages are migrated, the incentives will be against doing big design changes.

Reply:
This recorder also does more things than just being a httpClient.
Such as dealing with the start, stop, variables, and adding sanitizers.

How about renaming the type to `RecorderClient` and `RecorderClientCoreV1`? - intention is that RecorderClient does more than just being a HttpClient.
The same recorderClient(RecorderClientCoreV1) can be passed around as httpClient for core-v1 packages.

Reply:
I think that would improve things from where we are at the moment, but I wonder if we can do better. At the moment, I think the responsibilities of `TestProxyHttpClient` (with or without `CoreV1`) can be split into two areas:

1. The part which actually deals with the redirection of the requests (the HttpClient implementation for core v1 and the `configureClient` method for core v2); and
2. The interface that users of the recorder call to control (e.g. start and stop) the recorder.

The redirection part (1) that is used depends on whether the client library is core v1 or core v2, but the interface (2) is the same. If we could separate the two (e.g. by having a `httpClient` field as Daniel suggests), I feel like that's a more natural API.

On the other hand, as I said, that would require quite a big design change, so I am a bit on the fence. That said, if we're going to make changes, there really is no time like the present 🎁

Reply:
I really like those ideas @timovv ! 😱🔥